docs/concepts/what-is-a-workflow-definition.md

```diff
@@ -11,13 +11,97 @@ A Workflow Definition is the code that defines the constraints of a Workflow Exe
+The Event in the sequence must be an [ActivityTaskScheduled](/docs/concepts/what-is-an-event/#activitytaskscheduled) Event, where the Activity Name and the Task Queue name are the same as what is in the Command.
+
+If a Command is generated that doesn't match what it needs to in the existing Event History, then the Workflow Execution will return a "non-deterministic" error.
+
+The following are the two reasons why a Command might be generated out of sequence or the wrong Command might be generated altogether:
+
+1. Code changes are made to a Workflow Definition that is in use by a running Workflow Execution.
+2. There is intrinsic non-deterministic logic (i.e. inline random branching logic).
+
+#### Non-deterministic code changes
+
+The Workflow Definition can change in very limited ways once there is a Workflow Execution depending on it.
+To alleviate non-deterministic issues that arise from code changes we recommend using [Workflow Versioning](#what-is-workflow-versioning).
+
+For example, let's say we have a Workflow Definition that defines the following sequence:
+
+1. Start and wait on a Timer/sleep
+2. Spawn and wait on an Activity Execution
+3. Complete
+
+We start a Worker and spawn a Workflow Execution that uses that Workflow Definition.
+The Worker would emit the [StartTimer](/docs/concepts/what-is-a-command/#starttimer) Command and the Workflow Execution would become suspended.
+
+Before the Timer is up, we change the Workflow Definition to the following sequence:
+
+1. Spawn and wait on an Activity Execution
+2. Start and wait on a Timer/sleep
+3. Complete
+
+When the Timer fires, the next Workflow Task will cause the Workflow Function to re-execute.
+The first Command the Worker sees would be be ScheduleActivityTask Command, which wouldn't match up to the expected [TimerStarted](/docs/concepts/what-is-an-event#timerstarted] Event.
+
+The Workflow Execution would fail, and return the non-determinism error.
+
+The following are examples of some minor changes that would not take effect when re-executing a History which already contain the Events:
+
+- Changing the duration of a Timer.
+- Changing the arguments to:
+  - The Activity Options in a call to spawn an Activity Execution (local or nonlocal).
+  - The Child Workflow Options in a call to spawn a Child Workflow Execution.
+  - Call to Signal an External Workflow Execution.
+
+#### Intrinsic non-deterministic logic
+
+Intrinsic non-determinism is when a Workflow Function Execution might emit a different sequence of Commands on re-execution, regardless of whether all the input parameters are the same.
+
+For example, a Workflow Definition can not have inline logic that branches (emits a different Command sequence) based off a wall-clock or a random number.
+Here is some representative pseudocode:
+
+```text
+fn my_workflow() {
+  if system_clock().is_before("12pm") {
+    await workflow.sleep(duration_until("12pm"))
+  } else {
+    await my_afternoon_activity()
+  }
+}
+```
```

Comment:
The explanation about why `system_clock` is bad from here https://github.com/temporalio/documentation/pull/916/files#diff-02a12d0b950f980b1eaa3de8d4d05567f62c19673adb2315240d8655cb494ff4R123-R127

Is pretty important I think. Someone reading this document has to understand that `system_clock` is not a temporal API (which the name tries to make clear, but isn't immediately obvious). I think we need to re-enforce that and explain what should be used. The next paragraph explains that in generic terms, but I'm worried someone reading the code sample won't immediately connect why it's bad without something more specific.